### PR TITLE
[ENH] Refactor _loss_mapping 

### LINKED ISSUE #1957: [ENH] refactor `_loss_mapping` to tags of metrics, entirely
The `_loss_mapping` looks like it should be refactored to tags of metrics, entirely, in the respective `_pkg` classes.

Let´s do this! Though we should think about prioritization, or whether this is a "good first issue".

### PR DESCRIPTION
**Pull request by ParamThakkar123**
#### Reference Issues/PRs

Fixes #1957 

#### What does this implement/fix? Explain your changes.

This refactor moves legacy loss compatibility mappings and tensor shape checks to metric tags in the respective _pkg classes.
Test-specific parameters and dataloader configurations are now provided by a standardized _get_test_dataloaders_from method in each metric package.
All test logic for selecting compatible losses and checking output shapes is now tag-driven, improving maintainability and extensibility.
Legacy mapping code and if/else logic have been removed from the test suite.

#### PR checklist

- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [X] Added/modified tests
- [X] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

**Comment by ParamThakkar123** [comment]
@fkiraly

**Comment by ParamThakkar123** [comment]
Converting this to draft PR. Looking for feedback

**Comment by phoeenniixx** [comment]
If you have any problems/doubts, please feel free to ping us on discord. You can also join weekly community meet which happens every Friday at 13:00 UTC at the meetups channels on discord. 
https://discord.com/invite/54ACzaFsn7

**Comment by ParamThakkar123** [comment]
Hello @phoeenniixx Thanks for the amazing feedback. And yeah there are some metrics for which no compatible losses are found. So should I skip them in the test loop?

**Comment by ParamThakkar123** [comment]
Or what kind of changes should I exactly be doing ?

**Comment by ParamThakkar123** [comment]
A fallback?

**Comment by ParamThakkar123** [comment]
> Ahead of a review - you are reformatting a larger number of files. Could you kindly do that in a separate pull request, to avoid a large scope?

Yes. Will do that

**Comment by phoeenniixx** [comment]
> And yeah there are some metrics for which no compatible losses are found. 

So, here we use `metrics` as loss and, these compatible losses are calculated for the models which are being tested. These compatible losses are called in [test_all_estimators](https://github.com/sktime/pytorch-forecasting/blob/e1cc1ced27833741ad9896954785bfdae4f983f6/pytorch_forecasting/tests/test_all_estimators.py#L228) to pair the compatible losses with the models. So what is happening, imo, is right now `get_compatible_losses` is not collecting all the losses correctly.

> So should I skip them in the test loop?

These were collected before, so it should be possible to collect them with the new changes, so, I think we should not skip them. Rather, please check if there are some bugs

To check if all the losses are being collected correctly, please run tests from `TestAllPtForecasters` for both [v1](https://github.com/sktime/pytorch-forecasting/blob/e1cc1ced27833741ad9896954785bfdae4f983f6/pytorch_forecasting/tests/test_all_estimators.py#L370C7-L370C27) and [v2](https://github.com/sktime/pytorch-forecasting/blob/e1cc1ced27833741ad9896954785bfdae4f983f6/pytorch_forecasting/tests/test_all_estimators_v2.py#L83).

Ideally, both of these tests should pass and all the losses should be collected correctly.

**Comment by ParamThakkar123** [comment]
@phoeenniixx @fkiraly

**Comment by fkiraly** [comment]
@ParamThakkar123, how did you test whether the tests now get collected correctly?

**Comment by ParamThakkar123** [comment]
I ran the tests as mentioned by @phoeenniixx multiple times and looked at the errors that occurred then found bugs in my code which were related to some tags in the for loop for `LOSS_SPECIFIC_PARAMS`. Fixed them, ran tests locally again and pushed when all passed. Also while running the tests I added a print statement in the test code to check if anywhere `compatible_losses` is returned empty and it didn't after I made some more imports in the `_loss_mapping.py` actually the core issue was I wasn't importing some `_pkg`s which were actually required for certain losses. I made the imports, modified tags and ran the tests

@fkiraly

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py [hunk 1]
```diff
@@ -30,3 +40,25 @@ def get_encoder(cls):
         Returns a TorchNormalizer instance for rescaling parameters.
         """
         return TorchNormalizer(transformation="logit")
+
+    @classmethod
+    def _get_test_dataloaders_from(cls, params=None):
```
Comment (phoeenniixx) [inline pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: Can we not refactor it to a new position (like in a base class) from where it is called. As from what I can understand, there is a lot of duplication here.. A lot of `_pkg` classes use almost the similar code, so I think it would make sense to move it out as a standalone func (maybe to a base class)? and call it or override it in child classes if we need to customize it to a specific `_pkg` class? like we change the `target` to some other value for some losses, but this is not req for every loss. I was something like this
```python
# in the base class
@classmethod
def _get_test_dataloaders_from(cls, params=None, **kwargs):
       # code which is common for most of the metrics
       # kwargs are actually for the customisation purposes, 
       # like target="agency", sets target="agency".
```
```python
# package class
@classmethod
def _get_test_dataloders(cls, params=None):
      return _get_test_dataloaders_from(target="agency")
```
Does this makes sense? Please feel free to ask if I am not able to convey my thoughts clearly:)

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py [hunk 2]
```diff
@@ -16,6 +17,15 @@ class BetaDistributionLoss_pkg(_BasePtMetric):
         "distribution_type": "beta",
         "info:metric_name": "BetaDistributionLoss",
         "requires:data_type": "beta_distribution_forecast",
+        "clip_target": True,
```
Comment (phoeenniixx) [inline pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: I was wondering if we need `clip_target` and `data_loader_kwargs` in tags as they donot convey any useful info, rather they are the params to make the dataloaders more comaptible with each metric, maybe we should move it somewhere else in the class?

File: pytorch_forecasting/tests/_loss_mapping.py [hunk 3]
```diff
@@ -1,39 +1,52 @@
-from pytorch_forecasting.data.encoders import GroupNormalizer
 from pytorch_forecasting.metrics import (
-    MAE,
-    MAPE,
-    MASE,
-    RMSE,
-    SMAPE,
     BetaDistributionLoss,
-    CrossEntropy,
     ImplicitQuantileNetworkDistributionLoss,
```
Comment (phoeenniixx) [inline pytorch_forecasting/tests/_loss_mapping.py]: Do we need to import these loss classes? I think importing `_pkg` classes should suffice?

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py (hunk 1 above)
Comment (ParamThakkar123) [reply pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: Got it @phoeenniixx . Will implement this one 😄

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py (hunk 2 above)
Comment (ParamThakkar123) [reply pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: Yes. You're right. Where do you I should be moving it ?

File: pytorch_forecasting/tests/_loss_mapping.py (hunk 3 above)
Comment (ParamThakkar123) [reply pytorch_forecasting/tests/_loss_mapping.py]: Yeah. absolutely. Would make that change 😃

**Comment by ParamThakkar123** [comment]
@phoeenniixx I have implemented the changes you suggested. Can you please review it ?

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py (hunk 2 above)
Comment (phoeenniixx) [reply pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: maybe they can be one of the properties? or args in `__init__`? 
what do you think is a better idea? a `property` or an arg?

File: pytorch_forecasting/models/nhits/_nhits_pkg.py [hunk 4]
```diff
@@ -1,7 +1,25 @@
 """NHiTS package container."""
 
+from pytorch_forecasting.metrics.distributions import (
+    BetaDistributionLoss,
+    ImplicitQuantileNetworkDistributionLoss,
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/nhits/_nhits_pkg.py]: why are we importing these classes and `DISTR_LOSSES_NUMERIC` here in `_nhits_pkg`? are they usedf anywhere in the file?

File: pytorch_forecasting/models/nbeats/_nbeats_pkg.py [hunk 5]
```diff
@@ -40,6 +40,16 @@ def get_base_test_params(cls):
         """
         return [{}, {"backcast_loss_ratio": 1.0}]
 
+    @classmethod
+    def get_test_train_params(cls):
+        """
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/nbeats/_nbeats_pkg.py]: Please remove this function, I think `get_base_test_params` should be enough to get the params?

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py (hunk 2 above)
Comment (phoeenniixx) [reply pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: @fkiraly what is your opinion about this?

File: pytorch_forecasting/models/nhits/_nhits_pkg.py (hunk 4 above)
Comment (ParamThakkar123) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: @phoeenniixx I did that because there's a for loop in the code at line number 108 which loops over this `DISTR_LOSSES_NUMERIC` array

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py (hunk 2 above)
Comment (ParamThakkar123) [reply pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: @phoeenniixx I think we can keep it as an arg?

File: pytorch_forecasting/models/nhits/_nhits_pkg.py (hunk 4 above)
Comment (phoeenniixx) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: Hmm,  I see, but that still doesnt explain the imports of loss classes, from what i can see, they were already being imported in `_get_test_dataloaders_from` see [here](https://github.com/sktime/pytorch-forecasting/blob/e1cc1ced27833741ad9896954785bfdae4f983f6/pytorch_forecasting/models/nhits/_nhits_pkg.py#L75-L81)

Comment (phoeenniixx) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: Also, I am a bit uncomfortable about defining `DISTR_LOSSES_NUMERIC`, we imported it before as they were already present in `_loss_mapping`, and now you are refactoring it, so I think we could also refactor a little here? Like we just see if the loss is `MultivariateNormalDistributionLoss`, and we return `dataloaders_fixed_window_without_covariates()` and for rest of the distribution losses, we perform required operations and return `dataloaders_with_covariates`? idts we would need `DISTR_LOSSES_NUMERIC` then do we?

Comment (ParamThakkar123) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: Okay, So probably what we can do here is remove the redundant import but keep the `DISTR_LOSSES_NUMERIC ` as it is

Comment (phoeenniixx) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: I think we could even refactor `get_test_dataloaders_from` of `nhits_pkg` here as well :)
Please look at my previous comment. what do you think?

Also, why are we importing the unused classes here? (referring to your latest commit)

Comment (phoeenniixx) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: I think we dont even need to use `DISTR_LOSSES_NUMERIC` now, thanks to your brilliant refactoring :)

Comment (ParamThakkar123) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: Okay got it @phoeenniixx I will make the suggested changes

**Comment by ParamThakkar123** [comment]
@phoeenniixx I made the changes. Does the recent change look better now ?

File: pytorch_forecasting/models/nhits/_nhits_pkg.py (hunk 4 above)
Comment (phoeenniixx) [reply pytorch_forecasting/models/nhits/_nhits_pkg.py]: Just a heads up: there are other losses as well being tested for `nhits` along with distribution losses (like point predictions losses), so you'd have to add support for them as well, like previously there was for `TweedieLoss`(see the previous check for the loss) and for other losses `dataloaders_fixed_window_without_covariates` was being passed...

**Comment by phoeenniixx** [comment]
> @phoeenniixx I made the changes. Does the recent change look better now ?

Are all the tests passing locally?

**Comment by ParamThakkar123** [comment]
yup

File: pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py (hunk 2 above)
Comment (fkiraly) [reply pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py]: where is it being passed?

**Comment by ParamThakkar123** [comment]
@fkiraly They are being passed in the `_get_test_dataloaders_from` function

**Comment by fkiraly** [comment]
> They are being passed in the _get_test_dataloaders_from function

I see - that is a test config function, so I would either encode the parameters there, or in an additional private function - not in tags.

File: pytorch_forecasting/tests/_loss_mapping.py [hunk 6]
```diff
@@ -1,124 +1,84 @@
...
+
 
 def get_compatible_losses(pred_types, y_types):
-    """Get compatible losses based on prediction types and target types.
+    """
+    Get compatible losses based on prediction types and target types.
+    """
+    compatible_losses = []
+    for pkg in METRIC_PKGS:
+        pkg_pred_types = pkg._tags.get("info:pred_type", [])
```
Comment (phoeenniixx) [inline pytorch_forecasting/tests/_loss_mapping.py]: Here you could use `get_class_tags`, see [here](https://github.com/sktime/pytorch-forecasting/blob/e1cc1ced27833741ad9896954785bfdae4f983f6/pytorch_forecasting/tests/test_all_estimators.py#L144) for the example.

**Comment by phoeenniixx** [comment]
> > They are being passed in the _get_test_dataloaders_from function
> 
> I see - that is a test config function, so I would either encode the parameters there, or in an additional private function - not in tags.

Yes, that's why I proposed using `property` here... I think that could be a good way to access these params, and we could also add a way to change the value of this `property` if required, that would make it extensible as well?

**Comment by ParamThakkar123** [comment]
@phoeenniixx @fkiraly  added properties for clip_target and data_loader_kwargs and also replaces _tags.get with get_class_tags

**Comment by ParamThakkar123** [comment]
All tests pass locally

File: pytorch_forecasting/metrics/_distributions_pkg/_mqf2/_mqf2_distribution_loss_pkg.py [hunk 7]
```diff
@@ -18,6 +19,13 @@ class MQF2DistributionLoss_pkg(_BasePtMetric):
         "python_dependencies": ["cpflows"],
         "capability:quantile_generation": True,
         "requires:data_type": "mqf2_distribution_forecast",
+        "clip_target": True,
```
Comment (phoeenniixx) [inline pytorch_forecasting/metrics/_distributions_pkg/_mqf2/_mqf2_distribution_loss_pkg.py]: Add these to property as well? (all three `trainer_kwargs`, `data_loader_kwargs`, `clip_target`)

File: pytorch_forecasting/metrics/_distributions_pkg/_multivariate_normal/_multivariate_normal_distribution_loss_pkg.py [hunk 8]
```diff
@@ -17,12 +18,36 @@ class MultivariateNormalDistributionLoss_pkg(_BasePtMetric):
         "distribution_type": "multivariate_normal",
         "info:metric_name": "MultivariateNormalDistributionLoss",
         "requires:data_type": "multivariate_normal_distribution_forecast",
+        "info:pred_type": ["distr"],
+        "info:y_type": ["numeric"],
+        "loss_ndim": 2,
```
Comment (phoeenniixx) [inline pytorch_forecasting/metrics/_distributions_pkg/_multivariate_normal/_multivariate_normal_distribution_loss_pkg.py:23]: could you add these `loss_ndim` to all `metric_pkg` classes? Like I could see it is present in just distribution losses, but ig we could add it to other loss types as well? To point prediction losses and quantile loss?

File: pytorch_forecasting/models/nhits/_nhits_pkg.py [hunk 9]
```diff
@@ -111,15 +97,9 @@ def _get_test_dataloaders_from(cls, params):
 
             if isinstance(loss, NegativeBinomialDistributionLoss):
                 dwc = dwc.assign(volume=lambda x: x.volume.round())
-            # todo: still need some debugging to add the MQF2DistributionLoss
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/nhits/_nhits_pkg.py:114]: I'd suggest not to remove any "todo" until you are actually DOING the target specified there. They act like a pointer for the future, so that we dont forget the changes/improvements we are thinking of adding

File: pytorch_forecasting/models/nhits/_nhits_pkg.py [hunk 10]
```diff
@@ -51,28 +51,12 @@ def get_base_test_params(cls):
 
     @classmethod
     def _get_test_dataloaders_from(cls, params):
-        """Get dataloaders from parameters.
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/nhits/_nhits_pkg.py]: Donot remove the docstrings

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
Nice work! It is almost complete!
Just added some minor comments

**Comment by ParamThakkar123** [comment]
@phoeenniixx @fkiraly All changes implemented.

File: pytorch_forecasting/metrics/_quantile_pkg/_quantile_loss_pkg.py [hunk 11]
```diff
@@ -16,6 +16,9 @@ class QuantileLoss_pkg(_BasePtMetric):
         "metric_type": "quantile",
         "info:metric_name": "QuantileLoss",
         "requires:data_type": "quantile_forecast",
+        "info:pred_type": ["quantile"],
+        "info:y_type": ["numeric"],
+        "loss_ndim": 2,
```
Comment (phoeenniixx) [inline pytorch_forecasting/metrics/_quantile_pkg/_quantile_loss_pkg.py]: `loss_ndim` for `quantile_loss` would be `"num_quantiles"`. It would depend upon the number of quantiles that is being inputted by the user.

**Comment by phoeenniixx** [comment]
Thanks!
Very Nice! I think it is complete, just change the `loss_ndims`.

**Comment by ParamThakkar123** [comment]
Sure @phoeenniixx , noted. Making those changes in a minute

**Comment by phoeenniixx** [comment]
Actually I meant `"loss_ndim":"num_quantiles"`
I suggested to add `loss_ndims`  to tags in all `metrics` was because it is a piece of information that can be used to understand the loss and how it affects the output format of the models. 
So, what I was trying to say was in `QuantileLoss`, the `loss_ndims` is not fixed (like for other losses, you've seen). It depends on the number of quantiles being used. For example, if `num_quantiles = 3` with quantiles `[0.1, 0.5, 0.9]`, the loss is calculated separately for each quantile `(10%, 50%, and 90%)`, so `loss_ndims` would be 3

**Comment by ParamThakkar123** [comment]
> Actually I meant `"loss_ndim":"num_quantiles"` why I asked to add `loss_ndims` to tags in all `metrics` was because it is a piece of information that can be used to understand the loss and how it affects the output format of the models. So, what I was trying to say was in `QuantileLoss`, the `loss_ndims` is not fixed (like for other losses, you've seen). It depends on the number of quantiles being used. For example, if `num_quantiles = 3` with quantiles `[0.1, 0.5, 0.9]`, the loss is calculated separately for each quantile `(10%, 50%, and 90%)`, so `loss_ndims` would be 3

Got it. Really sorry for that

**Comment by phoeenniixx** [comment]
No worries, you are doing great!

**Comment by ParamThakkar123** [comment]
@phoeenniixx changes done. Can you please review again ? 😅

File: pytorch_forecasting/tests/_loss_mapping.py [hunk 12]
```diff
@@ -1,124 +1,82 @@
...
 }
 
 
 def get_compatible_losses(pred_types, y_types):
-    """Get compatible losses based on prediction types and target types.
+    """
+    Get compatible losses based on prediction types and target types.
+    """
+    compatible_losses = []
+    for pkg in METRIC_PKGS:
```
Comment (fkiraly) [inline pytorch_forecasting/tests/_loss_mapping.py:15]: instead of manually populating `METRIC_PKGS`, use `_registry.all_objects` to retrieve them

File: pytorch_forecasting/tests/_loss_mapping.py [hunk 13]
```diff
@@ -1,124 +1,82 @@
...
+    return pkg._get_test_dataloaders_from(params or {})
 
-    for pred_type in pred_types:
-        for y_type in y_types:
-            key = (pred_type, y_type)
-            if key in LOSSES_BY_PRED_AND_Y_TYPE:
-                compatible_losses.extend(LOSSES_BY_PRED_AND_Y_TYPE[key])
 
-    return compatible_losses
+def check_loss_output_shape(pkg, y_pred, y_true):
```
Comment (fkiraly) [inline pytorch_forecasting/tests/_loss_mapping.py]: question: is this function used anywhere? If not, did you forget to replace something with it?

File: pytorch_forecasting/models/nhits/_nhits_pkg.py [hunk 14]
```diff
@@ -84,19 +82,21 @@ def _get_test_dataloaders_from(cls, params):
...
-            if not isinstance(l, MultivariateNormalDistributionLoss)
-            # use dataloaders without covariates as default settings of nhits
-            # (hidden_size = 512) is not compatible with
-            # MultivariateNormalDistributionLoss causing Cholesky
-            # decomposition to fail during loss computation.
-        )
 
-        if isinstance(loss, distr_losses):
+        # Use fixed window dataloaders for MultivariateNormalDistributionLoss
+        if isinstance(loss, MultivariateNormalDistributionLoss):
```
Comment (fkiraly) [inline pytorch_forecasting/models/nhits/_nhits_pkg.py]: can this explicit checking now be replaced by checking only tags of the metrics?

If not, can we think about what tag could replace this?

File: pytorch_forecasting/tests/_loss_mapping.py (hunk 13 above)
Comment (ParamThakkar123) [reply pytorch_forecasting/tests/_loss_mapping.py]: Yes, actually I added that function just for checking the output shapes, its not used anywhere and I think we can remove it

**Comment by ParamThakkar123** [comment]
> * _get_test_dataloaders_from

The check_loss_output_shape is not being used anywhere just added for testing,removing it in the latest push and the _get_test_dataloaders_from is used in all the distribution, quantile and loss packages

**Comment by fkiraly** [comment]
can you say where precisely the methods are used, please?

File: pytorch_forecasting/tests/_loss_mapping.py (hunk 13 above)
Comment (fkiraly) [reply pytorch_forecasting/tests/_loss_mapping.py]: why did you add it? Or was it added by AI for no reason?

Comment (ParamThakkar123) [reply pytorch_forecasting/tests/_loss_mapping.py]: Yeah It was actually added while I was using copilot but it added it for checking shapes of tensors but it wasn't used anywhere later

**Comment by ParamThakkar123** [comment]
https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_distributions_pkg/_beta/_beta_distribution_loss_pkg.py#L57C24-L57C50

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_distributions_pkg/_implicit_quantile_network/_implicit_quantile_network_distribution_loss_pkg.py#L55

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_distributions_pkg/_log_normal/_log_normal_distribution_loss_pkg.py#L75

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_distributions_pkg/_multivariate_normal/_multivariate_normal_distribution_loss_pkg.py#L53

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_distributions_pkg/_negative_binomial/_negative_binomial_distribution_loss_pkg.py#L57

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_distributions_pkg/_normal/_normal_distribution_loss_pkg.py#L36

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_cross_entropy/_cross_entropy_pkg.py#L37

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_mae/_mae_pkg.py#L35

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_mape/_mape_pkg.py#L37

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_mase/_mase_pkg.py#L33

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_poisson/_poisson_loss_pkg.py#L37

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_rmse/_rmse_pkg.py#L35

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_smape/_smape_pkg.py#L37

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_point_pkg/_tweedie/_tweedie_loss_pkg.py#L35

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/_quantile_pkg/_quantile_loss_pkg.py#L46

Written here: 

https://github.com/ParamThakkar123/pytorch-forecasting/blob/d5765cadf1f7752ddb14668165827f9fd79c84a4/pytorch_forecasting/metrics/base_metrics/_base_object.py#L87

**Comment by ParamThakkar123** [comment]
@fkiraly After I replaced LOSS_SPECIFIC_PARAMS by loss_params there are errors from data_loader_kwargs but I made changes to those and ideally data_loader_kwargs should be passed to the dl_default_kwargs but they are not being passed. Is it because _get_test_dataloaders_from are differently accessed ?

**Comment by ParamThakkar123** [comment]
For distribution, point and quantile losses I am access _get_test_dataloaders_from from the base object while for other packages it's completely defined there as a class method. Does that require a refactor too @fkiraly ?

**Comment by ParamThakkar123** [comment]
@fkiraly should I keep `_get_test_dataloaders_from` only for the metric pkgs ?

**Comment by fkiraly** [comment]
I think we need to discuss this, I am a bit lost about the structure of the calls here. Do you have time to join the meeting about pytorch-forecasting on Sep 26, 13 UTC, on the sktime discord?

**Comment by ParamThakkar123** [comment]
How about 28th of September @fkiraly ?

**Comment by fkiraly** [comment]
works for me, but all the ptf developers are in the meeting on the 26th - may I recommend to get in touch on discord for scheduling?

**Comment by phoeenniixx** [comment]
Hi @ParamThakkar123, if you need any help regarding this PR, please reach us out at discord: https://discord.gg/3tVRa6UF
We can even have a discussion on this PR, test framework or `ptf`, `sktime` in general. 
We have community meet-ups every Friday at 13 UTC, feel free to join :)

**Comment by ParamThakkar123** [comment]
Sure @phoeenniixx !
